datafusion/core/src/execution/context.rs

```diff
@@ -1466,10 +1466,9 @@ impl SessionState {
         }
 
         let mut rules: Vec<Arc<dyn OptimizerRule + Sync + Send>> = vec![
-            // Simplify expressions first to maximize the chance
-            // of applying other optimizations
-            Arc::new(SimplifyExpressions::new()),
             Arc::new(PreCastLitInComparisonExpressions::new()),
```

Comment:
Now that we have `TypeCoercion` I wonder if we still need `PreCastLitInComparisonExpressions`

https://github.com/apache/arrow-datafusion/blob/d16457a0ba129b077935078e5cf89d028f598e0b/datafusion/optimizer/src/pre_cast_lit_in_comparison.rs#L31-L50

Appears to be a subset of what TypeCoercion is doing

Reply:
Though perhaps this is something similar to what you have described in https://github.com/apache/arrow-datafusion/issues/3622

Reply:
The `PreCastLitInComparisonExpressions` will be invalid when we move the `type coercion` to the front of `PreCastLitInComparisonExpressions`.

In the next pr, I will do https://github.com/apache/arrow-datafusion/issues/3622 and move the type coercion to the front of the `PreCastLitInComparisonExpressions`


I think `type coercion` just do one thing which is make type compatible in all operation, but the `PreCastLitInComparisonExpressions` or https://github.com/apache/arrow-datafusion/issues/3622 is to reduce `cast` for column expr or other expr instead of adding the `cast` to the literal. This will reduce the cast effort of the runtime.